util-images/network/netperfbenchmark/pkg/worker/worker.go

```diff
@@ -0,0 +1,284 @@
+		"/startTCPClient": w.StartTCPClient, "/startUDPServer": w.StartUDPServer, "/startUDPClient": w.StartUDPClient,
+		"/startHTTPServer": w.StartHTTPServer, "/startHTTPClient": w.StartHTTPClient, "/metrics": w.Metrics}
+	w.startListening(workerListenPort, h)
+	klog.Info("Started listening to Server")
+}
+
+func (w *Worker) startListening(port string, handlers map[string]func(http.ResponseWriter, *http.Request)) {
+	klog.Info("About to listen to controller requests")
+	for p, h := range handlers {
+		http.HandleFunc(p, h)
+	}
+	err := http.ListenAndServe(":"+port, nil)
+	if err != nil {
+		klog.Fatalf("Failed starting http server for port: %v, Error: %v", port, err)
+	}
+}
+
+// Metrics returns the metrics collected
+func (w *Worker) Metrics(res http.ResponseWriter, req *http.Request) {
+	var reply metricResponse
+	klog.Info("In metrics")
+	var stat string
+	select {
+	case stat = <-w.resultStatus:
+		if stat != "OK" {
+			klog.Error("Error collecting metrics:", stat)
+			reply.Error = "metrics collection failed:" + stat
+			w.createResponse(reply, &res)
+			return
+		}
+		klog.Info("Metrics collected")
+	default:
+		klog.Info("Metric collection in progress")
+		reply.Error = "metric collection in progress"
+		w.createResponse(reply, &res)
+		return
+	}
+	o, err := util.ParseResult(w.tstProtocol, w.result, w.tstDuration)
+	if err != nil {
+		reply.Error = err.Error()
+		w.createResponse(reply, &res)
+		return
+	}
+	reply.Result = o
+	//startime and other variables below can be removed in future, for debugging during initial runs
+	stStr := strconv.FormatInt(w.startedAt, 10)
+	ftStr := strconv.FormatInt(w.futureTime, 10)
+	dtStr := strconv.FormatInt(w.futureTime-w.startedAt, 10)
+	reply.WorkerStartTime = "StartedAt:" + stStr + " FutureTime:" + ftStr + " diffTime:" + dtStr
+	w.createResponse(reply, &res)
+}
+
+func (w *Worker) createResponse(resp interface{}, wr *http.ResponseWriter) {
+	klog.Info("Inside create resp")
+	(*wr).Header().Set("Content-Type", "application/json")
+	(*wr).WriteHeader(http.StatusOK)
+	klog.Info("Marshalled Resp:", resp)
+	b, err := json.Marshal(resp)
+	if err != nil {
+		klog.Info("Error marshalling to json:", err)
+	}
+	(*wr).Write(b)
+}
+
+// StartTCPClient starts iperf client for tcp measurements
+func (w *Worker) StartTCPClient(res http.ResponseWriter, req *http.Request) {
+	klog.Info("In StartTCPClient")
+	w.startWork(&res, req, util.ProtocolTCP, "iperf", tcpClientCmd)
+}
+
+// StartTCPServer starts iperf server for tcp measurements
+func (w *Worker) StartTCPServer(res http.ResponseWriter, req *http.Request) {
+	klog.Info("In StartTCPServer")
+	w.startWork(&res, req, util.ProtocolTCP, "iperf", tcpServerCmd)
+}
+
+// StartUDPServer starts iperf server for udp measurements
+func (w *Worker) StartUDPServer(res http.ResponseWriter, req *http.Request) {
+	klog.Info("In StartUDPServer")
+	w.startWork(&res, req, util.ProtocolUDP, "iperf", udpServerCmd)
+}
+
+// StartUDPClient starts iperf server for udp measurements
```

Comment:
Can you please sort them somehow (first HTTP, TCP, UDP, server before client or whatever)?

Reply:
Modified.